File: fixtures/rust/controller-service-provider/src/api/controller.rs
//! `PaymentController`: entry point of the refund flow.

use crate::provider::PaymentProvider;
use crate::service::payment_service::PaymentService;

/// HTTP-facing controller for payment endpoints.
pub struct PaymentController<P: PaymentProvider> {
    service: PaymentService<P>,
}

impl<P: PaymentProvider> PaymentController<P> {
    pub fn new(service: PaymentService<P>) -> Self {
        Self { service }
    }

    /// Handles `POST /refunds`.
    pub fn refund(&self, transaction_id: &str, amount_cents: u64) -> Result<String, String> {
        self.service.refund(transaction_id, amount_cents)
    }
}

File: fixtures/typescript/controller-service-provider/src/api/controller.ts
import { PaymentService, buildPaymentService } from "../service/paymentService";

export class PaymentController {
    private service: PaymentService;

    constructor() {
        this.service = buildPaymentService();
    }

    refund(amountCents: number): void {
        this.service.refund(amountCents);
    }
}
